### PR TITLE
React Hooks

### PR DESCRIPTION
**Pull request by excy-ai**
1. https://github.com/rolling-scopes-school/tasks/tree/master/react/modules/module03
2. imageimageimage
3. deploy not required for now
4. 30.03.2023 / 03.04.2023
5. self-assesment: 15 points
- [+] Eslint should not show any errors or warnings.
- [+] All the functionality should work.
- [+] Test coverage for the whole app should be not less than 60%.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -51,12 +53,46 @@
...
     "react": {
       "pragma": "React",
       "version": "detect"
+    },
+    "import/resolver": {
+      "typescript": true,
+      "node": true
     }
   }
 }
```
Comment (saha80) [inline .eslintrc:99]: No newline at end of file. You should run prettier on .json files

**Comment by saha80** [comment]
you should add ".husky" folder to gitignore

File: src/pages/FormPage/FormPage.tsx [hunk 2]
```diff
@@ -1,39 +1,33 @@
...
-            />
-          </aside>
-          <CardList data={this.state.cards} className="form-page__content" />
-        </div>
-      </>
-    );
-  }
-}
+  return (
+    <>
```
Comment (saha80) [inline src/pages/FormPage/FormPage.tsx]: redundant React.Fragment

File: src/components/select/SelectField.tsx [hunk 3]
```diff
@@ -1,28 +1,28 @@
...
-      </>
-    );
-  }
-}
+const SelectField: FC<SelectFieldProps> = (props) => {
+  const { error, options, register, label } = props;
+  const id = `select-${label}`;
+  const registerData = register ? register() : {};
+  return (
+    <>
```
Comment (saha80) [inline src/components/select/SelectField.tsx]: redundant React.Fragment

File: src/components/card/Card.tsx [hunk 4]
```diff
@@ -1,65 +1,64 @@
...
-          <CardField caption="Avg life span">{life_span}</CardField>
-          <CardField caption="Temperament">{temperament}</CardField>
-        </div>
-        {wikipedia_url && (
-          <Link target="_blank" to={wikipedia_url}>
-            WIKI
-          </Link>
-        )}
+  const {
+    reference_image_id,
```
Comment (saha80) [inline src/components/card/Card.tsx:16]: you should use lowerCamelCase for property
https://google.github.io/styleguide/tsguide.html#identifiers

File: src/utils/index.ts [hunk 5]
```diff
@@ -1,22 +1,5 @@
-import React from 'react';
 import { v4 as uuid } from 'uuid';
```
Comment (saha80) [inline src/utils/index.ts]:
```suggestion
export { v4 as generateId } from 'uuid';
```

File: src/pages/MainPage/MainPage.tsx [hunk 6]
```diff
@@ -1,34 +1,36 @@
...
+const MainPage: FC = () => {
+  const [query, setQuery] = useState(localStorage.getItem(QUERY_KEY) || '');
+  useEffect(() => {
+    return () => {
+      localStorage.setItem(QUERY_KEY, query);
+    };
+  });
+  return (
+    <main className="main">
+      <Search query={query} onChange={(query) => setQuery(query)} />
```
Comment (saha80) [inline src/pages/MainPage/MainPage.tsx]:
```suggestion
      <Search query={query} onChange={setQuery} />
```

File: src/components/layouts/base/BaseLayout.tsx [hunk 7]
```diff
@@ -1,20 +1,18 @@
...
-          <Outlet />
-        </main>
-      </React.Fragment>
-    );
-  }
-}
+const BaseLayout: FC = () => (
+  <>
+    <Header />
+    <main className={'main'}>
```
Comment (saha80) [inline src/components/layouts/base/BaseLayout.tsx:12]:
```suggestion
    <main className="main">
```

**Comment by saha80** [comment]
[repository requirements](https://github.com/rolling-scopes-school/tasks/blob/master/react/modules/module01/README.md#repository-requirements): **Do not merge Pull Request from the development branch to the main branch**

**Comment by excy-ai** [comment]
> [repository requirements](https://github.com/rolling-scopes-school/tasks/blob/master/react/modules/module01/README.md#repository-requirements): **Do not merge Pull Request from the development branch to the main branch**

Check announcements in RSS React Discord.
https://discord.com/channels/794806036506607647/1024014517711163503/1090029355197014046

**Comment by excy-ai** [comment]
> you should add ".husky" folder to gitignore

This is done on purpose, because it contains pre-commit hook. You shouldn't ignore it.

**Comment by saha80** [comment]
> > you should add ".husky" folder to gitignore
> 
> This is done on purpose, because it contains pre-commit hook. You shouldn't ignore it.

You're right. My bad.

File: src/components/card/Card.tsx [hunk 8]
```diff
@@ -1,65 +1,64 @@
...
+  } = props;
+  const flagUrl = `https://flagcdn.com/${country_code.toLowerCase()}.svg`;
+  const imageUrl = reference_image_id
+    ? `https://cdn2.thecatapi.com/images/${reference_image_id}.jpg`
+    : image_url;
+  const title = name.length > 25 ? name.substring(0, 25) : name;
+  return (
+    <div className="card">
+      <h3>{title}</h3>
+      <img className={'card__image'} src={imageUrl} onError={onImageError} alt={title} />
```
Comment (saha80) [inline src/components/card/Card.tsx:36]:
```suggestion
      <img className="card__image" src={imageUrl} onError={onImageError} alt={title} />
```

File: src/components/form/AddCardForm.tsx [hunk 9]
```diff
@@ -1,182 +1,148 @@
...
+              id={it}
+            />
+          ))}
+          {errors.temperament ? (
+            <span className="form__error">{errors.temperament.message}</span>
+          ) : (
+            errorPlaceholder
+          )}
+        </div>
+        <div className={'form__radio-group'}>
```
Comment (saha80) [inline src/components/form/AddCardForm.tsx:104]:
```suggestion
        <div className="form__radio-group">
```

File: src/components/header/Header.tsx [hunk 10]
```diff
@@ -1,29 +1,27 @@
...
-              <NavLink to={routes.form}>Form</NavLink>
-            </li>
-          </ul>
-        </nav>
-      </header>
-    );
-  }
-}
+const Header: FC = () => (
+  <header className={'header'}>
```
Comment (saha80) [inline src/components/header/Header.tsx:10]:
```suggestion
  <header className="header">
```

File: src/components/card/Card.tsx [hunk 11]
```diff
@@ -1,65 +1,64 @@
-import React, { ReactEventHandler } from 'react';
+import React, { FC, ReactEventHandler } from 'react';
```
Comment (saha80) [inline src/components/card/Card.tsx]: https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#removing-unused-react-imports
If you are using the new JSX transform from React 17, extend `"react/jsx-runtime"` in your eslint config (add `"plugin:react/jsx-runtime"` to `"extends"`) to disable the relevant rules.

File: src/components/form/AddCardForm.tsx [hunk 12]
```diff
@@ -1,182 +1,148 @@
...
-  };
+  const togglePopup = () => setIsPopupOpened(!isPopupOpened);
 
-  onSubmit = () => {
-    const files = this.fileRef.current?.files;
-    const file = files ? files[0] : undefined;
+  const onSubmit = (data: FormCard) => {
+    const { breed, date, temperament, lifespan, weight, image } = data;
+    const file = image ? image[0] : undefined;
     const imageUrl = file ? URL.createObjectURL(file) : undefined;
```
Comment (saha80) [inline src/components/form/AddCardForm.tsx:38]: used without `URL.revokeObjectURL`
https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL#memory_management

File: src/components/form/AddCardForm.tsx [hunk 13]
```diff
@@ -1,182 +1,148 @@
...
   };
 
-  clearForm = () => {
-    this.formRef.current?.reset();
-    this.togglePopup();
+  const clearForm = () => {
+    reset();
+    togglePopup();
   };
+  const required = { required: 'Field cannot be empty' };
```
Comment (saha80) [inline src/components/form/AddCardForm.tsx]: bad naming

File: src/components/input/InputField.tsx [hunk 14]
```diff
@@ -1,81 +1,78 @@
...
+      type={type}
+      accept={fileTypes}
+      defaultValue={value}
+      placeholder={label}
+    />
+  );
+  const errorComponent = error ? (
+    <span className="form__error">{error}</span>
+  ) : (
+    !isCheckable && <div className="form__error-placeholder" />
```
Comment (saha80) [inline src/components/input/InputField.tsx:42]: not readable nested condition in ternary operator

File: src/components/input/InputField.tsx [hunk 15]
```diff
@@ -1,81 +1,78 @@
...
+  );
+  const labelTextComponent = (label || value) && (
+    <span className={classNames('input-field__label', labelClassName)}>{label || value}</span>
+  );
+  const namedInput = isCheckable ? (
```
Comment (saha80) [inline src/components/input/InputField.tsx:47]: it should be a separate component

File: src/components/popup/Popup.tsx [hunk 16]
```diff
@@ -7,17 +8,13 @@ interface PopupProps {
...
-        style={{ zIndex: offset }}
-        className="popup__background"
-      >
-        <div className="popup">{this.props.children}</div>
-      </div>
-    );
-  }
-}
+const Popup: FC<PopupProps> = (props) => {
+  const offset = props.offset || 1;
```
Comment (saha80) [inline src/components/popup/Popup.tsx]: magic number

File: src/utils/index.ts (hunk 5 above)
Comment (saha80) [inline src/utils/index.ts]: consider using nanoid
https://github.com/ai/nanoid#comparison-with-uuid

Comment (excy-ai) [reply src/utils/index.ts]: Good point

Comment (excy-ai) [reply src/utils/index.ts]: As I see it wins on size of package, but in benchmark it is not. Thanks, but no.

File: src/components/card/Card.tsx (hunk 4 above)
Comment (excy-ai) [reply src/components/card/Card.tsx:16]: While it is imported from json - in my opinion it is ok. When you are working with external API do you ask them to rename it? Consider this as a result of external json loading for now.

File: src/components/card/Card.tsx (hunk 11 above)
Comment (excy-ai) [reply src/components/card/Card.tsx]: Thanks, will do

**Pull request merged**